File: android/src/main/kotlin/io/sentient/android/settings/SettingsDiagnostics.kt
// ---------------------------------------------------------------------------
// SettingsDiagnostics — the "Send diagnostic log" two-lane section of Settings.
//
// Lane 1: a button (settings-send-logs). Tap → reveals the session list.
// Lane 2: newest-first sessions, each human-labelled ("Today 9:43 PM"), crashed
//   ones flagged 🔴, "This session" (the newest) default-selected. Tapping a row's
//   send button MORPHS it in place into a progress bar (bound to progress:
//   Float?), then a result line ("Sent ✓ — ref XXXX" / "Upload failed — retry").
//
// Stateless: state (sessions / progress / outcome / which row is uploading) is
// hoisted from SettingsViewModel via the host. Split out of SettingsScreen to keep
// both files under the clean-code line limit.
// ---------------------------------------------------------------------------
package io.sentient.android.settings

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import io.sentient.android.theme.LocalTokens
import io.sentient.mobilesdk.design.Colors
import io.sentient.mobilesdk.vitals.VitalsSessionInfo

private const val SEND_LOGS_LABEL = "Send diagnostic log"
private const val DIAGNOSTICS_LABEL = "Diagnostics"
private const val CRASH_FLAG = "🔴 "
private const val ROW_TAG_PREFIX = "settings-log-session-"
private const val LABEL_SENT = "Sent ✓"
private const val SENT_REF_PREFIX = "Sent ✓ — ref "
private const val LABEL_RETRY = "Retry"
private const val LABEL_SEND = "Send"
private const val LABEL_SELECT = "Select"
private const val LABEL_NO_SESSIONS = "No diagnostic sessions yet."
private val PROGRESS_WIDTH = 64.dp

/**
 * The diagnostics section. [sessions] is newest-first; [progress] (0..1, null =
 * idle/done) and [outcome] reflect the in-flight / completed upload of [uploadingPath].
 */
@Composable
fun SettingsDiagnostics(
    sessions: List<VitalsSessionInfo>,
    nowMs: Long,
    uploadingPath: String?,
    progress: Float?,
    outcome: UploadOutcome?,
    onUpload: (String) -> Unit,
) {
    val tokens = LocalTokens.current
    var expanded by remember { mutableStateOf(false) }
    // Default-select the newest ("This session"); null only when there are no sessions.
    var selectedPath by remember(sessions) { mutableStateOf(sessions.firstOrNull()?.path) }

    Column(verticalArrangement = Arrangement.spacedBy(tokens.space.sm)) {
        Text(
            text = DIAGNOSTICS_LABEL,
            color = Color(Colors.ink3),
            fontSize = tokens.type.xs,
            fontWeight = FontWeight.SemiBold,
        )
        OutlinedButton(
            onClick = { expanded = !expanded },
            modifier = Modifier.fillMaxWidth().testTag("settings-send-logs"),
        ) { Text(SEND_LOGS_LABEL) }

        if (expanded) {
            if (sessions.isEmpty()) {
                Text(
                    text = LABEL_NO_SESSIONS,
                    modifier = Modifier.testTag("settings-log-empty"),
                    color = Color(Colors.ink3),
                    fontSize = tokens.type.sm,
                )
            } else {
                sessions.forEachIndexed { index, s ->
                    SessionUploadRow(
                        info = s,
                        label = sessionLabel(s, nowMs, isNewest = index == 0),
                        selected = s.path == selectedPath,
                        isUploading = s.path == uploadingPath,
                        progress = if (s.path == uploadingPath) progress else null,
                        outcome = if (s.path == uploadingPath) outcome else null,
                        onSelect = { selectedPath = s.path },
                        onUpload = { onUpload(s.path) },
                    )
                }
            }
        }
    }
}

/** One session: a select-row whose trailing send button morphs into a bar then a result. */
@Composable
private fun SessionUploadRow(
    info: VitalsSessionInfo,
    label: String,
    selected: Boolean,
    isUploading: Boolean,
    progress: Float?,
    outcome: UploadOutcome?,
    onSelect: () -> Unit,
    onUpload: () -> Unit,
) {
    val tokens = LocalTokens.current
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = tokens.space.xs)
            .testTag("$ROW_TAG_PREFIX${info.sessionStartMs}"),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(tokens.space.sm),
    ) {
        Text(
            text = (if (info.crashed) CRASH_FLAG else "") + label,
            modifier = Modifier.weight(1f),
            color = if (selected) Color(Colors.accent) else Color(Colors.ink),
            fontSize = tokens.type.sm,
            fontWeight = if (selected) FontWeight.SemiBold else FontWeight.Normal,
        )
        UploadControl(
            selected = selected,
            isUploading = isUploading,
            progress = progress,
            outcome = outcome,
            onSelect = onSelect,
            onUpload = onUpload,
        )
    }
}

/** The morphing trailing control: select → send button → progress bar → result line. */
@Composable
private fun UploadControl(
    selected: Boolean,
    isUploading: Boolean,
    progress: Float?,
    outcome: UploadOutcome?,
    onSelect: () -> Unit,
    onUpload: () -> Unit,
) {
    val tokens = LocalTokens.current
    when {
        isUploading && progress != null -> LinearProgressIndicator(
            progress = { progress },
            modifier = Modifier.width(PROGRESS_WIDTH).testTag("settings-log-progress"),
        )
        outcome is UploadOutcome.Sent -> Text(
            text = if (outcome.ref.isEmpty()) LABEL_SENT else "$SENT_REF_PREFIX${outcome.ref}",
            modifier = Modifier.testTag("settings-log-sent"),
            color = Color(Colors.accent),
            fontSize = tokens.type.xs,
        )
        outcome is UploadOutcome.Failed -> OutlinedButton(
            onClick = { onSelect(); onUpload() },
            modifier = Modifier.testTag("settings-log-failed"),
            colors = ButtonDefaults.outlinedButtonColors(contentColor = Color(Colors.stop)),
        ) { Text(LABEL_RETRY) }
        selected -> Button(
            onClick = onUpload,
            modifier = Modifier.testTag("settings-log-send"),
        ) { Text(LABEL_SEND) }
        else -> OutlinedButton(
            onClick = onSelect,
            modifier = Modifier.testTag("settings-log-select"),
        ) { Text(LABEL_SELECT) }
    }
}

File: ios/App/Settings/Diagnostics/SettingsDiagnostics.swift
// ---------------------------------------------------------------------------
// SettingsDiagnostics — the "Send diagnostic log" two-lane section of Settings.
// Swift mirror of Android's SettingsDiagnostics (settings/SettingsDiagnostics.kt).
//
// Lane 1: a button (settings-send-logs). Tap → reveals the session list.
// Lane 2: newest-first sessions, each human-labelled ("Today 9:43 PM"), crashed
//   ones flagged 🔴, "This session" (the newest) default-selected. Tapping a row's
//   send button MORPHS it in place into a progress bar (bound to progress: Double?),
//   then a result line ("Sent ✓ — ref XXXX" / "Retry").
//
// Reads the VM's published state; dispatches selection + upload as closures. The
// owning SettingsSheet builds + owns the SendLogsViewModel.
// ---------------------------------------------------------------------------
import SwiftUI
import MobileData

private let diagnosticsLabel = "Diagnostics"
private let sendLogsLabel = "Send diagnostic log"
private let crashFlag = "🔴 "
private let labelSent = "Sent ✓"
private let sentRefPrefix = "Sent ✓ — ref "
private let labelRetry = "Retry"
private let labelSend = "Send"
private let labelSelect = "Select"
private let labelNoSessions = "No diagnostic sessions yet."

/// The diagnostics section. Reads `model` published state; the newest session is
/// default-selected. Stateless beyond which row is selected (view-local).
struct SettingsDiagnostics: View {
    @ObservedObject var model: SendLogsViewModel
    let nowMs: Int64

    @State private var expanded = false
    @State private var selectedPath: String?

    var body: some View {
        VStack(alignment: .leading, spacing: Space.sm) {
            Text(diagnosticsLabel)
                .font(Typo.ui(TypeScale.base, .semibold))
                .foregroundStyle(DuskColors.ink3)

            sendLogsButton

            if expanded {
                switch model.loadPhase {
                case .loading:
                    DesignProgress(title: "Loading diagnostic sessions")
                        .accessibilityIdentifier("settings-log-loading")
                case .failed:
                    AsyncNotice(kind: .error, title: "Couldn't load diagnostic sessions") {
                        Task { await model.load() }
                    }
                case .ready where model.sessions.isEmpty:
                    AsyncNotice(kind: .empty, title: labelNoSessions)
                        .accessibilityIdentifier("settings-log-empty")
                case .ready:
                    sessionRows
                }
            }
        }
        .onChange(of: model.sessions) { _, sessions in
            // Default-select the newest ("This session") once the list loads.
            if selectedPath == nil { selectedPath = sessions.first?.path }
        }
    }

    private var sendLogsButton: some View {
        DesignActionButton(
            title: sendLogsLabel,
            role: .quiet,
            accessibilityId: "settings-send-logs",
            action: {
                expanded.toggle()
                if expanded && selectedPath == nil { selectedPath = model.sessions.first?.path }
            }
        )
    }

    private var sessionRows: some View {
        VStack(alignment: .leading, spacing: Space.xs) {
            ForEach(Array(model.sessions.enumerated()), id: \.element.path) { index, session in
                SessionUploadRow(
                    info: session,
                    label: SessionLabel.label(
                        sessionStartMs: session.sessionStartMs,
                        nowMs: nowMs,
                        isNewest: index == 0
                    ),
                    selected: session.path == selectedPath,
                    isUploading: session.path == model.uploadingPath,
                    progress: session.path == model.uploadingPath ? model.progress : nil,
                    outcome: session.path == model.uploadingPath ? model.outcome : nil,
                    onSelect: { selectedPath = session.path },
                    onUpload: { model.upload(path: session.path) }
                )
            }
        }
    }
}

#Preview("empty") {
    SettingsDiagnostics(
        model: SendLogsViewModel(initialLoadPhase: .ready),
        nowMs: Int64(Date().timeIntervalSince1970 * 1000)
    )
    .padding()
    .background(DuskColors.bg)
}

/// One session: a select-row whose trailing send control morphs into a bar then a result.
private struct SessionUploadRow: View {
    let info: VitalsSessionInfo
    let label: String
    let selected: Bool
    let isUploading: Bool
    let progress: Double?
    let outcome: UploadOutcome?
    let onSelect: () -> Void
    let onUpload: () -> Void

    var body: some View {
        HStack(spacing: Space.sm) {
            Text((info.crashed ? crashFlag : "") + label)
                .font(Typo.ui(TypeScale.sm, selected ? .semibold : .regular))
                .foregroundStyle(selected ? DuskColors.accent : DuskColors.ink)
                .frame(maxWidth: .infinity, alignment: .leading)
            UploadControl(
                selected: selected,
                isUploading: isUploading,
                progress: progress,
                outcome: outcome,
                onSelect: onSelect,
                onUpload: onUpload
            )
        }
        .padding(.vertical, Space.xs)
        .accessibilityValue(selected ? "Selected" : "Not selected")
        .accessibilityAddTraits(selected ? .isSelected : [])
        .accessibilityIdentifier("settings-log-session-\(info.sessionStartMs)")
    }
}

/// The morphing trailing control: select → send button → progress bar → result line.
private struct UploadControl: View {
    let selected: Bool
    let isUploading: Bool
    let progress: Double?
    let outcome: UploadOutcome?
    let onSelect: () -> Void
    let onUpload: () -> Void

    var body: some View {
        switch (isUploading, progress, outcome) {
        case let (true, p?, _):
            DesignProgress(value: p, accessibilityId: "settings-log-progress")
                .frame(width: DesignMetrics.progressWidth)
        case let (_, _, .sent(ref)):
            Text(ref.isEmpty ? labelSent : "\(sentRefPrefix)\(ref)")
                .font(Typo.ui(TypeScale.sm))
                .foregroundStyle(DuskColors.accent)
                .accessibilityIdentifier("settings-log-sent")
        case (_, _, .failed):
            DesignTextButton(
                title: labelRetry,
                role: .destructive,
                accessibilityId: "settings-log-failed",
                action: { onSelect(); onUpload() }
            )
        default:
            if selected {
                DesignTextButton(title: labelSend, role: .action, accessibilityId: "settings-log-send", action: onUpload)
            } else {
                DesignTextButton(title: labelSelect, accessibilityId: "settings-log-select", action: onSelect)
            }
        }
    }
}
